ulib/FStar.Matrix.fst

```diff
@@ -0,0 +1,839 @@
+
+(* This auxiliary lemma establishes the equality of the fold of the entire matrix
+   to the op of folds of (the submatrix of the first (m-1) rows) and (the last row). *) 
+let matrix_fold_snoc_lemma #c #eq 
+                           (#m: not_less_than 2) 
+                           (#n: pos) 
+                           (cm: CE.cm c eq) 
+                           (generator: matrix_generator c m n)
+  : Lemma (assert ((m-1)*n < m*n);
+            SP.foldm_snoc cm (matrix_seq generator) `eq.eq` 
+    cm.mult (SP.foldm_snoc cm (matrix_seq #c #(m-1) #n generator))
+            (SP.foldm_snoc cm (SB.slice (matrix_seq #c #m #n generator) ((m-1)*n) (m*n))))
+  = SB.lemma_eq_elim (matrix_seq generator)
+                  ((matrix_seq #c #(m-1) #n generator) `SB.append` 
+                  (SB.slice (matrix_seq generator) ((m-1)*n) (m*n)));    
+    SP.foldm_snoc_append cm (matrix_seq #c #(m-1) #n generator) 
+                         (SB.slice (matrix_seq generator) ((m-1)*n) (m*n)) 
+
+let foldm_snoc_last #c #eq (cm: CE.cm c eq) (s: SB.seq c{SB.length s > 0})
+  : Lemma (SP.foldm_snoc cm s == cm.mult (snd (SProp.un_snoc s)) 
+                                         (SP.foldm_snoc cm (fst (SProp.un_snoc s)))) 
+  = ()
+
+
+let matrix_submatrix_lemma #c (#m: not_less_than 2) (#n: pos)  
+                           (generator: matrix_generator c m n)
+  : Lemma ((matrix_seq generator) == (matrix_seq (fun (i:under(m-1)) (j:under n) -> generator i j) `SB.append` SB.init n (generator (m-1))))
+  = SB.lemma_eq_elim (matrix_seq (fun (i:under (m-1)) (j:under n) -> generator i j)) (matrix_seq #c #(m-1) #n generator);
+    SB.lemma_eq_elim (SB.slice (matrix_seq generator) ((m-1)*n) (m*n)) (SB.init n (generator (m-1)));
+    matrix_seq_decomposition_lemma generator
+
+(* This one could be probably written more efficiently -- 
+   but this implementation also works. *)
+#push-options "--ifuel 1 --fuel 0 --z3rlimit 130" 
+let rec matrix_fold_equals_fold_of_seq_folds #c #eq #m #n cm generator : Lemma 
+  (ensures foldm cm (init generator) `eq.eq`
+           SP.foldm_snoc cm (SB.init m (fun i -> SP.foldm_snoc cm (SB.init n (generator i)))) /\ 
+           SP.foldm_snoc cm (seq_of_matrix (init generator)) `eq.eq`
+           SP.foldm_snoc cm (SB.init m (fun i -> SP.foldm_snoc cm (SB.init n (generator i))))) 
+  (decreases m) =
+  let lhs_seq = matrix_seq generator in
+  let rhs_seq = SB.init m (fun i -> SP.foldm_snoc cm (SB.init n (generator i))) in
+  let lhs = SP.foldm_snoc cm (matrix_seq generator) in
+  let rhs = SP.foldm_snoc cm rhs_seq in
+  if m=1 then begin
+    SP.foldm_snoc_singleton cm (SP.foldm_snoc cm (SB.init n (generator 0)));
+    SB.lemma_eq_elim (SB.create 1 (SP.foldm_snoc cm (SB.init n (generator 0))))
+                  (SB.init m (fun i -> SP.foldm_snoc cm (SB.init n (generator i))));
+    SB.lemma_eq_elim (matrix_seq generator) (SB.init n (generator 0));
+    eq.symmetry rhs lhs 
+  end else 
+  let matrix = matrix_seq generator in 
```

Comment:
indentation

Reply:
The logic behind this is to only separate the terminal case from the main part, and keep the indentation minimal.

Morally I see it as being similar to not writing
```
let x = 1 in
  let y = 2 in
    let z = 3 in
```
but rather keeping it at the same level. 

If you think this is aesthetically worse, I'll fix that of course :)

Reply:
Yes, please fix it. This is not the convention used in the rest of the library.

Reply:
Done!